File: typescript/samples/GettingStartedGrpc/Checker.ts
import { BaseAgent } from "../../src/core/BaseAgent";
import { MessageContext } from "../../src/contracts/MessageContext";
import { CountMessage } from "./generated/message";
import { CountUpdate } from "./generated/message";

/**
 * Agent that checks and monitors a counter value.
 */
export class Checker extends BaseAgent {
    private count = 0;

    constructor() {
        super(
            { type: "checker", key: "default" },
            undefined as any, // Will be set by runtime
            "Agent that checks counter value"
        );
    }

    async handleAsync(message: unknown, context: MessageContext): Promise<unknown> {
        if (message instanceof CountUpdate) {
            this.count = message.newCount;
            console.log(`Checker: Count is now ${this.count}`);
            return null;
        }
        
        throw new Error("Unknown message type");
    }

    async startCount(): Promise<void> {
        await this.runtime.publishMessageAsync(
            new CountMessage({ count: 0 }),
            { type: "counter", source: "checker" }
        );
    }
}

File: dotnet/samples/GettingStartedGrpc/Checker.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Checker.cs

using Microsoft.AutoGen.Contracts;
using Microsoft.AutoGen.Core;
using Microsoft.Extensions.Hosting;
using TerminationF = System.Func<int, bool>;

namespace GettingStartedGrpcSample;

[TypeSubscription("default")]
public class Checker(
    AgentId id,
    IAgentRuntime runtime,
    IHostApplicationLifetime hostApplicationLifetime,
    TerminationF runUntilFunc
    ) :
        BaseAgent(id, runtime, "Modifier", null),
        IHandle<Events.CountUpdate>
{
    public async ValueTask HandleAsync(Events.CountUpdate item, MessageContext messageContext)
    {
        if (!runUntilFunc(item.NewCount))
        {
            Console.WriteLine($"\nChecker:\n{item.NewCount} passed the check, continue.");
            await this.PublishMessageAsync(new Events.CountMessage { Content = item.NewCount }, new TopicId("default"));
        }
        else
        {
            Console.WriteLine($"\nChecker:\n{item.NewCount} failed the check, stopping.");
            hostApplicationLifetime.StopApplication();
        }
    }
}
